ludwig/utils/defaults.py

```diff
@@ -16,357 +16,38 @@
-        full_input_feature_config, _ = load_config_with_kwargs(
-            input_type_registry[input_feature[TYPE]].get_schema_cls(), input_feature
-        )
-        input_feature.update(asdict(full_input_feature_config))
-
-    # ===== Combiner =====
-    set_default_value(config, COMBINER, {TYPE: default_combiner_type})
-    full_combiner_config, _ = load_config_with_kwargs(
-        combiner_registry[config[COMBINER][TYPE]].get_schema_cls(), config[COMBINER]
-    )
-    config[COMBINER].update(asdict(full_combiner_config))
-
-    # ===== Output features =====
-    for output_feature in config[OUTPUT_FEATURES]:
-        if config[MODEL_TYPE] == MODEL_GBM:
-            set_default_values(output_feature, {DECODER: {TYPE: "passthrough"}})
-            remove_ecd_params(output_feature)
-        get_from_registry(output_feature[TYPE], output_type_registry).populate_defaults(output_feature)
-
-        # By default, drop rows with missing output features
-        set_default_value(output_feature, PREPROCESSING, {})
-        set_default_value(output_feature[PREPROCESSING], "missing_value_strategy", DROP_ROW)
-
-        # Update decoder and loss related parameters for output feature from global defaults
-        apply_global_defaults(config, output_feature, OUTPUT_FEATURES)
-
-        # Express all output feature parameters back into the config.
-        full_output_feature_config, _ = load_config_with_kwargs(
-            output_type_registry[output_feature[TYPE]].get_schema_cls(), output_feature
-        )
-        output_feature.update(asdict(full_output_feature_config))
-
-    # ===== Hyperpot =====
-    if HYPEROPT in config:
-        set_default_value(config[HYPEROPT][EXECUTOR], TYPE, RAY)
-    return config
-
-
-def remove_ecd_params(feature):
-    feature.pop("tied", None)
-    feature.pop("fc_layers", None)
-    feature.pop("num_layers", None)
-    feature.pop("output_size", None)
-    feature.pop("use_bias", None)
-    feature.pop("weights_initializer", None)
-    feature.pop("bias_initializer", None)
-    feature.pop("norm", None)
-    feature.pop("norm_params", None)
-    feature.pop("activation", None)
-    feature.pop("dropout", None)
-    feature.pop("embedding_size", None)
-    feature.pop("embeddings_on_cpu", None)
-    feature.pop("pretrained_embeddings", None)
-    feature.pop("embeddings_trainable", None)
-    feature.pop("embedding_initializer", None)
-    # decoder params
-    feature.pop("reduce_input", None)
-    feature.pop("dependencies", None)
-    feature.pop("reduce_dependencies", None)
-    feature.pop("loss", None)
-    feature.pop("num_fc_layers", None)
-    feature.pop("threshold", None)
-    feature.pop("clip", None)
-    feature.pop("top_k", None)
+default_preprocessing_parameters.update(PreprocessingConfig().to_dict())
 
 
 def render_config(config=None, output=None, **kwargs):
-    output_config = merge_with_defaults(config)
+    upgraded_config = upgrade_to_latest_version(config)
```

Comment:
I think this is how it should work in general, where `update_to_latest_version` and from_dict are separate operations.
Just pointing this out because `update_to_latest_version` also called inside Config.__init__